File: sdks/typescript/test/client.test.ts
import { createHmac } from "node:crypto";
import { describe, expect, it, vi } from "vitest";

import {
  AuthenticationError,
  Forgefy,
  JobFailedError,
  JobTimeoutError,
  QuotaExceededError,
  RateLimitError,
  ValidationError,
  verifySignature,
} from "../src/index.js";

const BASE = "https://api.example.com";

function jsonResponse(status: number, body: unknown, headers: Record<string, string> = {}): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "Content-Type": "application/json", ...headers },
  });
}

function problem(status: number, detail: string): Response {
  return jsonResponse(status, { type: "about:blank", title: "err", status, detail });
}

const EXTRACT_OK = {
  id: "e1",
  model_tier: "standard",
  features: [{ title: "OAuth login", description: "SSO", priority: "high" }],
  questions: [],
  conflicts: [],
  action_items: [],
  usage: { input_tokens: 100, output_tokens: 40 },
};

function makeClient(fetchMock: typeof fetch, opts: Partial<ConstructorParameters<typeof Forgefy>[0]> = {}) {
  return new Forgefy({
    apiKey: "fgy_live_test",
    baseUrl: BASE,
    fetch: fetchMock,
    retryDelayMs: 1, // keep retry tests fast
    ...opts,
  });
}

describe("extract", () => {
  it("sends the key, body, and parses the result", async () => {
    const fetchMock = vi.fn(async () => jsonResponse(200, EXTRACT_OK));
    const client = makeClient(fetchMock as unknown as typeof fetch);

    const result = await client.extract({
      transcript: "we need oauth",
      extractors: ["features"],
      modelTier: "standard",
    });

    expect(result.features[0].title).toBe("OAuth login");
    const [url, init] = fetchMock.mock.calls[0] as unknown as [string, RequestInit];
    expect(url).toBe(`${BASE}/api/v1/extract`);
    expect((init.headers as Record<string, string>).Authorization).toBe("Bearer fgy_live_test");
    expect(JSON.parse(init.body as string)).toEqual({
      transcript: "we need oauth",
      extractors: ["features"],
      model_tier: "standard",
    });
  });

  it("omits optional fields it was not given", async () => {
    const fetchMock = vi.fn(async () => jsonResponse(200, EXTRACT_OK));
    const client = makeClient(fetchMock as unknown as typeof fetch);

    await client.extract({ transcript: "hello world" });

    const [, init] = fetchMock.mock.calls[0] as unknown as [string, RequestInit];
    expect(JSON.parse(init.body as string)).toEqual({ transcript: "hello world" });
  });

  it("maps problem+json to typed errors", async () => {
    const cases: Array<[Response, unknown]> = [
      [problem(401, "bad key"), AuthenticationError],
      [problem(402, "quota exhausted, resets on August 1"), QuotaExceededError],
      [problem(422, "transcript too long"), ValidationError],
    ];
    for (const [response, errorClass] of cases) {
      const client = makeClient(vi.fn(async () => response) as unknown as typeof fetch);
      await expect(client.extract({ transcript: "t" })).rejects.toBeInstanceOf(errorClass as never);
    }
  });

  it("retries 429 then succeeds", async () => {
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce(problem(429, "slow down"))
      .mockResolvedValueOnce(jsonResponse(200, EXTRACT_OK));
    const client = makeClient(fetchMock as unknown as typeof fetch);

    const result = await client.extract({ transcript: "t" });
    expect(result.id).toBe("e1");
    expect(fetchMock).toHaveBeenCalledTimes(2);
  });

  it("gives up with RateLimitError after maxRetries", async () => {
    const fetchMock = vi.fn(async () => problem(429, "slow down"));
    const client = makeClient(fetchMock as unknown as typeof fetch, { maxRetries: 1 });

    await expect(client.extract({ transcript: "t" })).rejects.toBeInstanceOf(RateLimitError);
    expect(fetchMock).toHaveBeenCalledTimes(2); // initial + 1 retry
  });

  it("does NOT retry the sync extract on 5xx (tokens would double-bill)", async () => {
    const fetchMock = vi.fn(async () => problem(502, "all extractors errored"));
    const client = makeClient(fetchMock as unknown as typeof fetch);

    await expect(client.extract({ transcript: "t" })).rejects.toMatchObject({ status: 502 });
    expect(fetchMock).toHaveBeenCalledTimes(1);
  });
});

describe("jobs", () => {
  it("create auto-generates an Idempotency-Key and retries 5xx safely", async () => {
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce(problem(502, "queue hiccup"))
      .mockResolvedValueOnce(jsonResponse(202, { job_id: "j1", status: "queued", webhook_secret: null }));
    const client = makeClient(fetchMock as unknown as typeof fetch);

    const job = await client.jobs.create({ transcript: "long one" });

    expect(job.job_id).toBe("j1");
    expect(fetchMock).toHaveBeenCalledTimes(2);
    const keys = fetchMock.mock.calls.map(
      (c) => ((c as unknown as [string, RequestInit])[1].headers as Record<string, string>)["Idempotency-Key"],
    );
    expect(keys[0]).toBeTruthy();
    expect(keys[0]).toBe(keys[1]); // the retry reuses the same key — no double job
  });

  it("create passes an explicit idempotency key and webhook url through", async () => {
    const fetchMock = vi.fn(async () =>
      jsonResponse(202, { job_id: "j1", status: "queued", webhook_secret: "s3cret" }),
    );
    const client = makeClient(fetchMock as unknown as typeof fetch);

    const job = await client.jobs.create({
      transcript: "t",
      webhookUrl: "https://example.com/hook",
      idempotencyKey: "import-42",
    });

    expect(job.webhook_secret).toBe("s3cret");
    const [, init] = fetchMock.mock.calls[0] as unknown as [string, RequestInit];
    expect((init.headers as Record<string, string>)["Idempotency-Key"]).toBe("import-42");
    expect(JSON.parse(init.body as string).webhook_url).toBe("https://example.com/hook");
  });

  it("waitFor polls until done", async () => {
    const statuses = ["queued", "processing", "done"];
    let call = 0;
    const fetchMock = vi.fn(async () =>
      jsonResponse(200, {
        job_id: "j1",
        status: statuses[call++],
        model_tier: "standard",
        created_at: "2026-07-16T00:00:00Z",
        result: call === 3 ? { features: [], usage: { input_tokens: 1, output_tokens: 1 } } : null,
        error: null,
      }),
    );
    const client = makeClient(fetchMock as unknown as typeof fetch);

    const job = await client.jobs.waitFor("j1", { pollIntervalMs: 1 });
    expect(job.status).toBe("done");
    expect(fetchMock).toHaveBeenCalledTimes(3);
  });

  it("waitFor throws JobFailedError on failed", async () => {
    const fetchMock = vi.fn(async () =>
      jsonResponse(200, {
        job_id: "j1",
        status: "failed",
        model_tier: "standard",
        created_at: "2026-07-16T00:00:00Z",
        result: null,
        error: "provider down",
      }),
    );
    const client = makeClient(fetchMock as unknown as typeof fetch);

    await expect(client.jobs.waitFor("j1", { pollIntervalMs: 1 })).rejects.toBeInstanceOf(JobFailedError);
  });

  it("waitFor times out", async () => {
    const fetchMock = vi.fn(async () =>
      jsonResponse(200, {
        job_id: "j1",
        status: "processing",
        model_tier: "standard",
        created_at: "2026-07-16T00:00:00Z",
        result: null,
        error: null,
      }),
    );
    const client = makeClient(fetchMock as unknown as typeof fetch);

    await expect(client.jobs.waitFor("j1", { pollIntervalMs: 1, timeoutMs: 5 })).rejects.toBeInstanceOf(
      JobTimeoutError,
    );
  });
});

describe("usage", () => {
  it("fetches the quota snapshot", async () => {
    const fetchMock = vi.fn(async () =>
      jsonResponse(200, {
        tier: "starter",
        tier_name: "Starter",
        monthly_tokens: 5_000_000,
        tokens_used: 1_200_000,
        tokens_remaining: 3_800_000,
        resets_at: "2026-08-01T00:00:00+00:00",
      }),
    );
    const client = makeClient(fetchMock as unknown as typeof fetch);

    const usage = await client.usage();
    expect(usage.tokens_remaining).toBe(3_800_000);
    const [url] = fetchMock.mock.calls[0] as unknown as [string];
    expect(url).toBe(`${BASE}/api/v1/usage`);
  });
});

describe("verifySignature", () => {
  const secret = "s3cret";
  const body = JSON.stringify({ type: "extract.job.completed", job_id: "j1" });
  const goodSig = `sha256=${createHmac("sha256", secret).update(body).digest("hex")}`;

  it("accepts a valid signature", () => {
    expect(verifySignature(body, goodSig, secret)).toBe(true);
  });

  it("accepts the bare hex form too", () => {
    expect(verifySignature(body, goodSig.slice(7), secret)).toBe(true);
  });

  it("rejects a tampered body", () => {
    expect(verifySignature(body + "x", goodSig, secret)).toBe(false);
  });

  it("rejects the wrong secret", () => {
    expect(verifySignature(body, goodSig, "other")).toBe(false);
  });

  it("rejects garbage headers without throwing", () => {
    expect(verifySignature(body, undefined, secret)).toBe(false);
    expect(verifySignature(body, "", secret)).toBe(false);
    expect(verifySignature(body, "sha256=nothex", secret)).toBe(false);
  });
});

File: sdks/dart/test/client_test.dart
import 'dart:convert';

import 'package:forgefy/forgefy.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

const _tinyDelay = Duration(milliseconds: 1);

Forgefy _client(MockClient mock) => Forgefy(
      apiKey: 'fgy_test_123',
      baseUrl: 'https://api.example.test',
      httpClient: mock,
      retryDelay: _tinyDelay,
    );

void main() {
  group('extract', () {
    test('sends auth + body and parses the result', () async {
      late http.Request captured;
      final mock = MockClient((req) async {
        captured = req;
        return http.Response(
          jsonEncode({
            'id': 'abc',
            'model_tier': 'standard',
            'features': [
              {'title': 'Google login', 'description': 'OAuth', 'priority': 'high'}
            ],
            'questions': [],
            'conflicts': [],
            'action_items': [],
            'usage': {'input_tokens': 10, 'output_tokens': 20},
          }),
          200,
          headers: {'content-type': 'application/json'},
        );
      });

      final forgefy = _client(mock);
      final result = await forgefy.extract(
        transcript: 'We need Google login',
        extractors: [Extractor.features, Extractor.actionItems],
      );

      expect(captured.method, 'POST');
      expect(captured.url.path, '/api/v1/extract');
      expect(captured.headers['authorization'], 'Bearer fgy_test_123');
      final sent = jsonDecode(captured.body) as Map<String, dynamic>;
      expect(sent['extractors'], ['features', 'action_items']);

      expect(result.id, 'abc');
      expect(result.features.single.title, 'Google login');
      expect(result.usage.inputTokens, 10);
    });

    test('retries on 429 then succeeds', () async {
      var calls = 0;
      final mock = MockClient((req) async {
        calls++;
        if (calls == 1) return http.Response('{"detail":"slow down"}', 429);
        return http.Response(
          jsonEncode({
            'id': 'x',
            'model_tier': 'standard',
            'features': [],
            'questions': [],
            'conflicts': [],
            'action_items': [],
            'usage': {'input_tokens': 0, 'output_tokens': 0},
          }),
          200,
        );
      });

      final result = await _client(mock).extract(transcript: 'hi there friend');
      expect(calls, 2);
      expect(result.id, 'x');
    });

    test('maps 401 to AuthenticationException', () async {
      final mock = MockClient(
        (req) async => http.Response('{"detail":"bad key"}', 401),
      );
      expect(
        () => _client(mock).extract(transcript: 'hello world here'),
        throwsA(isA<AuthenticationException>()
            .having((e) => e.status, 'status', 401)
            .having((e) => e.detail, 'detail', 'bad key')),
      );
    });
  });

  group('jobs', () {
    test('create sends an Idempotency-Key', () async {
      late http.Request captured;
      final mock = MockClient((req) async {
        captured = req;
        return http.Response(
          jsonEncode(
              {'job_id': 'job-1', 'status': 'queued', 'webhook_secret': 'sek'}),
          202,
        );
      });

      final job = await _client(mock).jobs.create(transcript: 'long transcript here');
      expect(captured.headers['idempotency-key'], isNotEmpty);
      expect(job.jobId, 'job-1');
      expect(job.status, JobStatus.queued);
      expect(job.webhookSecret, 'sek');
    });

    test('waitFor polls until done', () async {
      var calls = 0;
      final mock = MockClient((req) async {
        calls++;
        final status = calls < 2 ? 'processing' : 'done';
        return http.Response(
          jsonEncode({
            'job_id': 'job-2',
            'status': status,
            'model_tier': 'standard',
            'created_at': '2026-01-01T00:00:00Z',
            'result': status == 'done' ? {'features': []} : null,
            'error': null,
          }),
          200,
        );
      });

      final job = await _client(mock)
          .jobs
          .waitFor('job-2', pollInterval: _tinyDelay);
      expect(job.status, JobStatus.done);
      expect(calls, 2);
    });

    test('waitFor throws JobFailedException on failure', () async {
      final mock = MockClient(
        (req) async => http.Response(
          jsonEncode({
            'job_id': 'job-3',
            'status': 'failed',
            'model_tier': 'standard',
            'created_at': '2026-01-01T00:00:00Z',
            'result': null,
            'error': 'all extractors errored',
          }),
          200,
        ),
      );
      expect(
        () => _client(mock).jobs.waitFor('job-3', pollInterval: _tinyDelay),
        throwsA(isA<JobFailedException>()
            .having((e) => e.jobId, 'jobId', 'job-3')),
      );
    });
  });

  group('verifySignature', () {
    // HMAC-SHA256 of 'hello' keyed by 'secret', hex.
    const body = 'hello';
    const secret = 'secret';
    const validHex =
        '88aab3ede8d3adf94d26ab90d3bafd4a2083070c3bcce9c014ee04a443847c0b';

    test('accepts a valid sha256= signature', () {
      expect(verifySignature(body, 'sha256=$validHex', secret), isTrue);
    });

    test('accepts a bare hex signature', () {
      expect(verifySignature(body, validHex, secret), isTrue);
    });

    test('rejects a wrong signature, secret, or missing header', () {
      expect(verifySignature(body, 'sha256=${'0' * 64}', secret), isFalse);
      expect(verifySignature(body, 'sha256=$validHex', 'wrong'), isFalse);
      expect(verifySignature(body, null, secret), isFalse);
      expect(verifySignature(body, 'sha256=notarealhash', secret), isFalse);
    });
  });
}
